### PR TITLE
Entity tags

### PR DESCRIPTION
**Pull request by emina**
*Issue #, if available:*

[RFC 82](https://github.com/cedar-policy/rfcs/blob/emina/entity-tags/text/0082-entity-tags.md)

*Description of changes:*

This is a draft PR that adds entity tags to the Cedar Lean model and proofs.  It makes the following changes:

* Model entity tag operations (`.hasTag` and  `.getTag`) in the evaluator and validator. 
* Update the JSON parsing code by making guesses about the JSON format for entity tags. This code may need to be revised, depending on the format we adopt for encoding tag information.
* Remove `Partial` model and theorems for now (to be updated later in a separate branch).
* Update validator proofs.

### CODE REVIEW

File: cedar-lean/Cedar/Validation/Typechecker.lean [hunk 1]
```diff
@@ -157,16 +166,36 @@ def typeOfInₛ (ety₁ ety₂ : EntityType) (x₁ x₂ : Expr) (env : Environme
...
   | some _ => ok (.bool .anyBool)
   | none   => err (.lubErr ty₁ ty₂)
 
-def typeOfBinaryApp (op₂ : BinaryOp) (ty₁ ty₂ : CedarType) (x₁ x₂ : Expr) (env : Environment) : ResultType :=
+def typeOfBinaryApp (op₂ : BinaryOp) (ty₁ ty₂ : CedarType) (x₁ x₂ : Expr) (c : Capabilities) (env : Environment) : ResultType :=
   match op₂, ty₁, ty₂ with
   | .eq, _, _                               => typeOfEq ty₁ ty₂ x₁ x₂
   | .mem, .entity ety₁, .entity ety₂        => ok (.bool (typeOfInₑ ety₁ ety₂ x₁ x₂ env))
   | .mem, .entity ety₁, .set (.entity ety₂) => ok (.bool (typeOfInₛ ety₁ ety₂ x₁ x₂ env))
+  | .hasTag, .entity ety₁, .string          => typeOfHasTag ety₁ x₁ x₂ c env
```
Comment (aaronjeline) [inline cedar-lean/Cedar/Validation/Typechecker.lean:197]: Is there a reason we have `GetAttr`/`HasAttr` as separate AST notes but not for tags?

Comment (aaronjeline) [reply cedar-lean/Cedar/Validation/Typechecker.lean:197]: Ah it's that the rhs is an expression not an `Attr`, that makes sense

Comment (emina) [reply cedar-lean/Cedar/Validation/Typechecker.lean:197]: Yes, that was the rationale.  Curious: did we choose to make them separate nodes in Rust?

File: cedar-lean/Cedar/Spec/Evaluator.lean [hunk 2]
```diff
@@ -57,6 +57,8 @@ def apply₂ (op₂ : BinaryOp) (v₁ v₂ : Value) (es : Entities) : Result Val
   | .containsAny, .set vs₁, .set vs₂                       => .ok (vs₁.intersects vs₂)
   | .mem, .prim (.entityUID uid₁), .prim (.entityUID uid₂) => .ok (inₑ uid₁ uid₂ es)
   | .mem, .prim (.entityUID uid₁), .set (vs)               => inₛ uid₁ vs es
+  | .hasTag, .prim (.entityUID uid₁), .prim (.string tag)  => .ok ((es.tagsOrEmpty uid₁).contains tag)
+  | .getTag, .prim (.entityUID uid₁), .prim (.string tag)  => (es.tagsOrEmpty uid₁).findOrErr tag .tagDoesNotExist
```
[the remark this answers is not in the document]
Comment (emina) [reply cedar-lean/Cedar/Spec/Evaluator.lean:61]: Getting a tag on an entity that's not in store will be an error.  

Note that `es.tagsOrEmpty uid₁` will return the empty map if `uid₁` is not in the store.  Then, `findOrErr` will error when we attempt to look up `tag` in the empty map. 

That said, the error thrown (`tagDoesNotExist`) will be different than the error thrown for `getAttr` (`entityDoesNotExist`), so if we want those to match, it's easy to change this code to get the matching behavior.

Can do in the next rev.

File: cedar-lean/Cedar/Validation/Typechecker.lean (hunk 1 above)
Comment (cdisselkoen) [reply cedar-lean/Cedar/Validation/Typechecker.lean:197]: In Rust I added them to `BinaryOp` as well

**Review by aaronjeline — APPROVED** [review]

**Review by shaobo-he-aws — APPROVED** [review]

**Pull request merged**